### PR TITLE
1965 add credentials UI form

### LINKED ISSUE #1965: Decouple credentials and configuration
## Description

Managing propagation credentials presents an interesting challenge. Some credentials are configured, while others are stolen. Currently, configured credentials are part of the configuration, and stolen credentials get added to the configuration. This introduces complexity, since the island has to keep track of per-agent configurations and initial configs vs. the config that gets modified during runtime.

Configuration should be static while agents are running! As part of the agent refactor project, the agent now ignores the credentials in the configuration and requests propagation credentials from a separate endpoint. From the agent's point of view, it doesn't care if credentials were stolen or configured; it just needs to use them.

Decouple credentials and configuration. Allow the user to configure credentials as always, but the UI should submit them to a separate endpoint from configuration. Implement the `ICredentialsRepository` to store collected credentials in MongoDB. The `ICredentialsRepository` should make it simple for the caller to retrieve all credentials, just those credentials that were stolen (for example, by mimikatz), or just those credentials that the user configured. Modify the existing telemetry processor and retrieval endpoint to store and retrieve credentials from the `ICredentialsRepository`. Modify configuration import/export to have a new format containing the "config" and configured credentials.

## Tasks
- [x] Modify `IStolenCredentialsRepository` (0d) @ilija-lazoroski 
    - [x] Rename to `ICredentialsRepository`
    - [x] Add `{get,save}_configured_credentials()`
    - [x] Add `get_all_credentials()`
    - [x] Add docstrings
        - [x] Return appropriate `{Retrieval,Storage,Removal}Errors`
    - [x] Add `remove_{configured,stolen,all}_credentials()`
- [x] Credential serialization (0d) - @mssalvatore 
- [x] Implement a `MongoCredentialsRepository` that stores credentials (0d) @ilija-lazoroski 
    - [x] Modify the interface to accept a sequence of credentials when storing
- [x] Encrypt/decrypt the secrets in the `MongoCredentialsRepository (0d) @mssalvatore 
- [x] Add `/api/propagation-credentials/stolen` (0d) @VakarisZ 
    - [x] Define GET
    - [x] Define POST
    - [x] Define DELETE
- [x] Add `/api/propagation-credentials/configured` (0d) @VakarisZ 
    - [x] Define GET
    - [x] Define POST
    - [x] Define DELETE
- [x] Add `GET /api/propagation-credentials` that returns all types of credentials ie two collections (0d) @VakarisZ 
- [x] Add `DELETE /api/propagation-credentials` that removes two sub-collections: stolen and configured credentials (0d) @VakarisZ 
- [x] Modify credentials telemetry processor to use `ICredentialsRepository` (0d) @ilija-lazoroski 
- [x] Modify the UI to submit usernames, passwords, NTLM hashes, and SSH keys to `/api/propagation-credentials/configured` (0.25d) @VakarisZ 
    - [x] Add a new form that allows the user to add the above credentials
    - [x] Add the form to the "Propagation" tab, or a new tab, whichever seems better
- [x] Modify reports to get credentials from `GET /api/propagation-credentials` (0d) - @ilija-lazoroski 
    - [x] Modify the backend report to not include credentials
    - [x] Modify the UI to get credentials from the API and add them to the report
- [x] Modify `/api/clear-simulation-data` to clear stolen credentials (0d) - @shreyamalviya 
- [x] ~Add `/api/reset-configured-propagation-credentials` (0.0d) - @shreyamalviya~
    - [x] ~Call from the UI when configuration is reset~
- [x] Use DELETE /api/propagation-credentials/configured in the UI when configuration is reset (0d) - @shreyamalviya

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Adds propagation credentials to the UI.
#1965

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] Added relevant unit tests?
* [ ] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [ ] If applicable, add screenshots or log transcripts of the feature working

image
image
image

### CODE REVIEW

File: monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  const [displayedSchema, setDisplayedSchema] = useState(getSchemaByKey(schema, initialSection));
+  const [displayedSchemaUi, setDisplayedSchemaUi] = useState(getUiSchemaByKey(uiSchema, initialSection));
+  const onInnerDataChange = (innerData) => {
+    console.log(innerData);
+    formData[selectedSection] = innerData.formData;
+    onChange({formData: formData});
+  }
+
+  const setSection = (sectionKey) => {
+    console.log(sectionKey);
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js:35]: Missed log

File: monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    customFormats,
+    className,
+    formData
+  } = props;
+  console.log(schema, uiSchema, formData);
+  const [selectedSection, setSelectedSection] = useState(initialSection);
+  const [displayedSchema, setDisplayedSchema] = useState(getSchemaByKey(schema, initialSection));
+  const [displayedSchemaUi, setDisplayedSchemaUi] = useState(getUiSchemaByKey(uiSchema, initialSection));
+  const onInnerDataChange = (innerData) => {
+    console.log(innerData);
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js:29]: missed log

File: monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      {sectionOrder.map(section => {
+        return (
+          <Nav.Item key={section}>
+            <Nav.Link eventKey={section}>{getNavTitle(schema, section)}</Nav.Link>
+          </Nav.Item>);
+      })}
+    </Nav>)
+  }
+
+  console.log(displayedSchemaUi)
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js:56]: missed log

File: monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import React, {useState} from 'react';
+import {Nav} from 'react-bootstrap';
+import {CREDENTIALS} from '../../services/configuration/credentials.js';
+
+const sectionOrder = [
+  'exploitation',
+  'network_scan',
+  'credentials',
+  'maximum_depth'
+];
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js:11]: Why are tabs defined here and not in configuration page?

Comment (VakarisZ) [reply monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js:11]: We already have a method `getSectionsOrder` in ConfigurePage.js

File: monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          onChange={onInnerDataChange}
+          customFormats={customFormats}
+          className={className}
+          liveValidate>
+      <button type='submit' className={'hidden'}>Submit</button>
+    </Form>
+  </div>)
+}
+
+function getSchemaByKey(schema, key) {
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js:72]: Why can't we just have the schema in this format, why do we need to dynamically create it?

Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/components/configuration-components/PropagationConfig.js:72]: Maximum_depth is not an object so we can't return properties of it. Maybe you can jump in and help me with a bunch of these stuff..

**Review by VakarisZ — COMMENTED** [review]
Nice work, but we still need to fetch credentials on load and put them into the configuration. On submit we need to send them to a separate endpoint and save there.

File: monkey/monkey_island/cc/ui/src/components/pages/ConfigurePage.js [hunk 6]
```diff
@@ -302,13 +303,18 @@ class ConfigurePageComponent extends AuthComponent {
...
-    return (
-      <div>
-        <Form {...formProperties} key={displayedSchema.title}>
-          <button type='submit' className={'hidden'}>Submit</button>
-        </Form>
-      </div>
-    )
+    if (this.state.selectedSection === 'propagation') {
+        console.log(formProperties['uiSchema']);
+        return (<PropagationConfig {...formProperties}/>)
```
[the remark this answers is not in the document]
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/components/pages/ConfigurePage.js:308]: We are not passing the whole schema, just the one from propagation.

File: monkey/monkey_island/cc/ui/src/services/configuration/propagation/credentials.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  'title': 'Credentials',
+  'type': 'object',
+  'properties': {
+    'exploit_user_list': {
+      'title': 'Exploit user list',
+      'type': 'array',
+      'uniqueItems': true,
+      'items': {'type': 'string'},
+      'default': [],
+      'description': 'List of user names that will be used by exploiters that need ' +
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/ui/src/services/configuration/propagation/credentials.js:11]:
```suggestion
      'description': 'List of usernames that will be used by exploiters that need ' +
```

**Review by mssalvatore — COMMENTED** [review]
Looks good, but I don't see where the credentials are submitted to the island. Did I just miss it?

**Pull request merged**